kale/evaluate/similarity_metrics.py

```diff
@@ -0,0 +1,158 @@
+"""
+Authors: Lawrence Schobs, lawrenceschobs@gmail.com
+
+Functions related to similarity metrics including similarity measures and correlations.
+"""
+
+import logging
+import os
+from typing import Any, Dict, List, Optional, Tuple
+
+import numpy as np
+import pandas as pd
+
+from kale.prepdata.tabular_transform import apply_confidence_inversion
+
+
+def jaccard_similarity(list1: list, list2: list) -> float:
+    """
+    Calculates the Jaccard Index (JI) between two lists.
+    Args:
+        list1 (list): List of elements in set A.
+        list2 (list): List of elements in set B.
+    Returns:
+        float: JI between list1 and list2.
+    Raises:
+        None.
+    Example:
+        >>> jaccard_similarity([1,2,3], [2,3,4])
+        0.5
+    """
+
+    if len(list1) == 0 or len(list2) == 0:
+        return 0
+    else:
+        intersection = len(set(list1).intersection(list2))  # no need to call list here
+        union = len(list1 + list2) - intersection  # you only need to call len once here
+        return intersection / union  # also no need to cast to float as this will be done for you
+
+
+def evaluate_correlations(
+    bin_predictions: Dict[str, pd.DataFrame],
+    uncertainty_error_pairs: List[Tuple[str, str, str]],
+    cmaps: List[Dict[Any, Any]],
+    num_bins: int,
+    conf_invert_info: List[Tuple[str, bool]],
+    num_folds: int = 8,
+    error_scaling_factor: float = 1,
+    combine_middle_bins: bool = False,
+    save_path: Optional[str] = None,
+    to_log: bool = False,
+) -> Dict[str, Dict[str, Dict[str, Any]]]:
+    """
+    Calculates the correlation between error and uncertainty for each bin and for each target,
+    using a piece-wise linear regression model.
+
+    Designed for use in Quantile Binning (/pykale/examples/landmark_uncertainty/main.py).
+
+    Args:
+        bin_predictions: A dictionary of Pandas DataFrames containing model predictions for each testing fold.
+        uncertainty_error_pairs: A list of tuples specifying the names of the uncertainty,
+            error, and uncertainty inversion keys for each pair.
+        cmaps: A dictionary of colour maps to use for plotting the results.
+        num_bins: The number of quantile bins to divide the data into.
+        conf_invert_info: A list of tuples specifying whether to invert the uncertainty values for each method.
```

Comment:
what `conf_` stands for? Why not say list? What type for each element in the list? Not boolean? The name says info.